### PR TITLE
fix(cron): various follow-ups for multiple schedules

### PR DESCRIPTION
**Pull request by eduardodbr**
### Motivation

Currently `argo cron list` and `argo cron get` are showing the schedules with timezone, when available. This PR removes it as it doesn't bring any value because it can be seen in the `timezone` field.

### Modifications

The schedule string can now be requested with and without the timezone. 
With timezone, it is used for the annotation `last-used-schedule`, so changing the timezone while keeping the schedule will be seen as a new schedule. 
Without timezone, it is used in `argo cron` CLI.

Also:
- update every instance of `schedule` to be `schedules` as suggested by @agilgur5 https://github.com/argoproj/argo-workflows/pull/12616#discussion_r1501850996
- update fields in CronWorkflows table for multiple schedules per https://github.com/argoproj/argo-workflows/pull/12616#pullrequestreview-1889629343
- add in-line version references and deprecation per https://github.com/argoproj/argo-workflows/pull/12616#discussion_r1495330587
- add example

### Verification

Current: 

`argo cron list` 

image

`argo cron get`

image

With this PR changes:

`argo cron list`

image

`argo cron get`

image

The `CronWorkflow` label `last-used-schedule` still uses the schedule with timezone

image

### CODE REVIEW

**Review by agilgur5 — COMMENTED** [review]
It might make sense to split out the `schedule` -> `schedules` change https://github.com/argoproj/argo-workflows/pull/12616#discussion_r1501850996 into its own PR. That and the `v3.6 and after:` and other #12616 review follow-ups can be merged pretty much immediately while you're still working out the test failures etc here. And the `v3.6 and after:` portion is particularly useful for people who are looking at the `latest` docs.

Also there was one miss from my previous review https://github.com/argoproj/argo-workflows/pull/12616#pullrequestreview-1889629343, the CronWorkflow docs page's fields table

**Comment by eduardodbr** [comment]
whille implementing this, I also found that `schedule` was still marked as required in the open api spec/json schema/CRD spec so I removed it. Do you think it should have a dedicated PR or is it ok to have it here? @agilgur5

**Comment by agilgur5** [comment]
Is the CLI fix only relevant to 3.6? Or does it also impact 3.5/3.4?

If these are all just follow-up fixes to the multiple `schedules` PR, we can leave it all as one I suppose -- just would rename the PR.

If the CLI fix is relevant to 3.5/3.4, then I think we should split the rest out, so that it can be backported

**Comment by eduardodbr** [comment]
> Is the CLI fix only relevant to 3.6? Or does it also impact 3.5/3.4?
> 
> If these are all just follow-up fixes to the multiple `schedules` PR, we can leave it all as one I suppose -- just would rename the PR.
> 
> If the CLI fix is relevant to 3.5/3.4, then I think we should split the rest out, so that it can be backported

this only applies to 3.6 as it was introduced in #12616

**Comment by agilgur5** [comment]
I know `schedules` was, but I wasn't sure if this TZ duplication happened in earlier versions with single `schedule` as well

**Comment by eduardodbr** [comment]
> I know `schedules` was, but I wasn't sure if this TZ duplication happened in earlier versions with single `schedule` as well

the timezone in `argo cron lilst` and `argo cron get` was also introduced in that PR

**Comment by agilgur5** [comment]
Ok gotcha, wanted to make sure since I hadn't reviewed that whole PR. Then let's just fix-up the title and description to include all follow-ups as one

Also btw, no need to quote reply if you're replying to the whole thing in order. Quoting is useful for replying to parts at a time or out of order comments. In this case they read fluidly down and so quoting actually reads duplicatively

File: ui/src/app/cron-workflows/cron-workflow-list.tsx [hunk 1]
```diff
@@ -156,7 +156,7 @@ export function CronWorkflowList({match, location, history}: RouteComponentProps
                                         <div className='columns small-2'>{w.metadata.namespace}</div>
                                         <div className='columns small-1'>{w.spec.timezone}</div>
                                         <div className='columns small-1'>
-                                            {w.spec.schedule != ''
+                                            {(w.spec.schedule ?? '') != ''
```
Comment (agilgur5) [inline ui/src/app/cron-workflows/cron-workflow-list.tsx:159]:
```suggestion
                                            {w.spec.schedule
```

you can just use [truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) for all these

File: ui/src/app/cron-workflows/cron-workflow-list.tsx [hunk 2]
```diff
@@ -156,7 +156,7 @@ export function CronWorkflowList({match, location, history}: RouteComponentProps
...
                                                 ? w.spec.schedule
                                                 : w.spec.schedules.map(schedule => (
```
Comment (agilgur5) [inline ui/src/app/cron-workflows/cron-workflow-list.tsx:161]:
```suggestion
                                            {w.spec.schedule || w.spec.schedules.map(schedule => (
```

this specific one can be further simplified

Comment (agilgur5) [reply ui/src/app/cron-workflows/cron-workflow-list.tsx:161]: although tbh normalizing `schedule` to a single element array of `schedules` would be simpler to do everywhere

**Comment by Joibel** [comment]
@eduardodbr, any chance that you'd have a chance to fix/update this so it can go into 3.6?

We're currently missing the documentation on multiple `schedules`, which is in this PR.

**Comment by eduardodbr** [comment]
Hi @Joibel will try to finish this ASAP, sorry for the delay

File: docs/cron-workflows.md [hunk 3]
```diff
@@ -43,7 +43,8 @@ You can use `CronWorkflow.spec.workflowMetadata` to add `labels` and `annotation
 
 | Option Name                  | Default Value          | Description |
 |:----------------------------:|:----------------------:|-------------|
-| `schedule`                   | None, must be provided | [Cron schedule](#cron-schedule-syntax) to run `Workflows`. Example: `5 4 * * *` |
+| `schedule`                   | None | [Cron schedule](#cron-schedule-syntax) to run `Workflows`. Example: `5 4 * * *`. Either `schedule` or `schedules` must be provided. |
```
Comment (agilgur5) [inline docs/cron-workflows.md:46]:
```suggestion
| `schedule`                   | None | [Cron schedule](#cron-schedule-syntax) to run `Workflows`. Example: `5 4 * * *`. Deprecated, see `schedules`. |
```

We can probably modify this to just say that `schedule` is deprecated. Similar to #13358.

@Joibel does that make sense to you for this feature?

File: pkg/apis/workflow/v1alpha1/cron_workflow_types.go [hunk 4]
```diff
@@ -45,7 +45,7 @@ type CronWorkflowSpec struct {
 	// WorkflowSpec is the spec of the workflow to be run
 	WorkflowSpec WorkflowSpec `json:"workflowSpec" protobuf:"bytes,1,opt,name=workflowSpec,casttype=WorkflowSpec"`
 	// Schedule is a schedule to run the Workflow in Cron format
```
Comment (agilgur5) [inline pkg/apis/workflow/v1alpha1/cron_workflow_types.go:47]:
```suggestion
	// Schedule is a schedule to run the Workflow in Cron format. Deprecated, see Schedules
```

same as above

File: examples/cron-workflow-multiple-schedules.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+apiVersion: argoproj.io/v1alpha1
+kind: CronWorkflow
+metadata:
+  name: hello-world-multiple-schedules
+spec:
+  schedules: 
+  - "*/3 * * * *"
+  - "*/2 * * * *"
```
Comment (agilgur5) [inline examples/cron-workflow-multiple-schedules.yaml:8]:
```suggestion
  schedules: # v3.6 and after
    - "*/3 * * * *"
    - "*/2 * * * *"
```

- add in-line version reference here
- consistent indentation

**Review by agilgur5 — COMMENTED** [review]
some notes on in-line deprecation and versions

File: docs/cron-workflows.md (hunk 3 above)
Comment (Joibel) [reply docs/cron-workflows.md:46]: Yes, I agree.
It is a trivial change to upgrade and totally automatable. We can then remove the old implementation sooner.

File: ui/src/app/cron-workflows/cron-workflow-list.tsx [hunk 6]
```diff
@@ -222,9 +222,10 @@ export function CronWorkflowList({match, location, history}: RouteComponentProps
     );
 }
 
-function getCronNextScheduledTime(spec: CronWorkflowSpec): Date {
-    if (spec.schedule != '') {
-        return getNextScheduledTime(spec.schedule, spec.timezone);
+function getSpecNextScheduledTime(spec: CronWorkflowSpec): Date {
+    if (spec.schedule && spec.schedule.trim() !== '') {
+        const nextTime = getNextScheduledTime(spec.schedule, spec.timezone);
+        return nextTime;
```
Comment (agilgur5) [inline ui/src/app/cron-workflows/cron-workflow-list.tsx:228]:
```suggestion
        return getNextScheduledTime(spec.schedule, spec.timezone);
```

unnecessary variable

Comment (agilgur5) [reply ui/src/app/cron-workflows/cron-workflow-list.tsx:228]: I'm also not sure we should have the `trim()` check here and nowhere else. I don't even think the UI should handle that tbh, and I'm not sure that would be valid YAML either

**Review by agilgur5 — COMMENTED** [review]
Thanks for updating this! Just a small comment below

File: ui/src/app/cron-workflows/cron-workflow-list.tsx (hunk 6 above)
Comment (eduardodbr) [reply ui/src/app/cron-workflows/cron-workflow-list.tsx]: sorry, I found an issue with next run time when using multiple schedules and I changed the code to debug. I forgot to revert. Should be good now

File: ui/src/app/cron-workflows/cron-workflow-list.tsx [hunk 7]
```diff
@@ -222,8 +222,8 @@ export function CronWorkflowList({match, location, history}: RouteComponentProps
     );
 }
 
-function getCronNextScheduledTime(spec: CronWorkflowSpec): Date {
-    if (spec.schedule != '') {
+function getSpecNextScheduledTime(spec: CronWorkflowSpec): Date {
+    if (spec.schedule && spec.schedule !== '') {
```
Comment (agilgur5) [inline ui/src/app/cron-workflows/cron-workflow-list.tsx:226]:
```suggestion
    if (spec.schedule) {
```

truthy covers the case of empty string already, so the `&& ..` is redundant

**Review by agilgur5 — APPROVED** [review]
LGTM, thanks for iterating on this!

**Comment by agilgur5** [comment]
Looks like there's some test failures in CI

**Pull request merged**